broker/client/client.go

```diff
@@ -712,3 +514,276 @@ func getPeerInfo(peer *ill_db.Peer, symbol string) (string, iso18626.TypeAgencyI
+
+func (c *Iso18626Client) determineMessageTarget(ctx extctx.ExtendedContext, trCtx transactionContext) (*messageTarget, error) {
+	msgSupplierSymbol := getSymbol(trCtx.event.EventData.IncomingMessage)
+	isNotification := isNotification(trCtx.event.EventData.IncomingMessage)
+	if trCtx.selectedSupplier == nil {
+		return c.handleNoSelectedSupplier(ctx, trCtx, msgSupplierSymbol, isNotification)
+	}
+
+	if isNotification && msgSupplierSymbol != "" && msgSupplierSymbol != trCtx.selectedSupplier.SupplierSymbol {
+		return c.handleSkippedSupplierNotification(ctx, trCtx, msgSupplierSymbol)
+	}
+
+	return handleSelectedSupplier(trCtx)
+}
+
+func (c *Iso18626Client) handleNoSelectedSupplier(ctx extctx.ExtendedContext, trCtx transactionContext, msgSupplierSymbol string, isNotification bool) (*messageTarget, error) {
+	if isNotification && msgSupplierSymbol != "" {
+		return c.handleSkippedSupplierNotification(ctx, trCtx, msgSupplierSymbol)
+	}
+
+	return &messageTarget{status: iso18626.TypeStatusUnfilled, firstMessage: false}, nil
+}
+
+func (c *Iso18626Client) handleSkippedSupplierNotification(ctx extctx.ExtendedContext, trCtx transactionContext, msgSupplierSymbol string) (*messageTarget, error) {
+	skipped, skippedPeer, err := c.getSkippedSupplierAndPeer(ctx, trCtx.transaction.ID, msgSupplierSymbol)
+	if err != nil {
+		return nil, err
+	}
+
+	if trCtx.requester.BrokerMode == string(extctx.BrokerModeTransparent) {
+		return &messageTarget{
+			supplier:     skipped,
+			peer:         skippedPeer,
+			status:       iso18626.TypeStatus(trCtx.transaction.LastSupplierStatus.String),
+			firstMessage: false,
+		}, nil
+	}
+
+	return &messageTarget{
+		problemDetails: &struct{ message string }{
+			message: fmt.Sprintf("ignored notification from skipped supplier %s due to requester mode %s", msgSupplierSymbol, trCtx.requester.BrokerMode),
+		},
+	}, nil
+}
+
+func handleSelectedSupplier(trCtx transactionContext) (*messageTarget, error) {
+	lastReceivedStatus := trCtx.selectedSupplier.LastStatus
+	if s, ok := iso18626.StatusMap[lastReceivedStatus.String]; ok {
+		return &messageTarget{supplier: trCtx.selectedSupplier, peer: trCtx.selectedPeer, status: s, firstMessage: false}, nil
+	}
+
+	if !lastReceivedStatus.Valid {
+		if trCtx.event.EventData.IncomingMessage == nil {
+			if trCtx.requester.BrokerMode == string(extctx.BrokerModeTransparent) || trCtx.requester.BrokerMode == string(extctx.BrokerModeTranslucent) {
+				return &messageTarget{supplier: trCtx.selectedSupplier, status: BrokerInfoStatus, firstMessage: true}, nil
+			}
+			return &messageTarget{note: fmt.Sprintf(BrokerDoesNotSendInThisMode, BrokerInfoStatus, trCtx.requester.BrokerMode)}, nil
+		}
+		return &messageTarget{supplier: trCtx.selectedSupplier, status: BrokerInfoStatus, firstMessage: true}, nil
+	}
+
+	return nil, fmt.Errorf(FailedToResolveStatus, lastReceivedStatus.String)
+}
+
+func buildSupplyingAgencyMessage(trCtx transactionContext, target *messageTarget) *iso18626.ISO18626Message {
```

Comment:
@JanisSaldabols rename this function to `create...` or the `createRequest....` functions to `build..` for consistency.